src/ESPeasyGPIO.ino

```diff
@@ -0,0 +1,393 @@
+// c) Does not update the pinstate state if the port is defined in a task, as it will be picked up by thePLUGIN_TEN_PER_SECOND event
+// called from: run10TimesPerSecond in ESPEasy.ino
+//*********************************************************
+
+void GPIO_Monitor10xSec()
+{
+  for (auto it=globalMapPortStatus.begin(); it!=globalMapPortStatus.end(); ++it) {
+    //only call monitor function if there is the need to
+    if (it->second.monitor || it->second.command || it->second.init) {
+      const uint16_t gpioPort = getPortFromKey(it->first);
+      const uint16_t pluginID = getPluginFromKey(it->first);
+      int8_t currentState;
+      String eventString;
+      bool caseFound = true;
+
+      switch (pluginID)
+      {
+        case PLUGIN_GPIO:
+          currentState = GPIO_Read_Switch_State(gpioPort, it->second.mode);
+          eventString = F("GPIO");
+          break;
+        case PLUGIN_MCP:
+          currentState = GPIO_MCP_Read(gpioPort);
+          if (currentState==-1){
+            it->second.mode=PIN_MODE_OFFLINE;
+          } else if (it->second.mode==PIN_MODE_OFFLINE) {
+            it->second.mode=PIN_MODE_UNDEFINED;
+          }
+          eventString = F("MCP");
+          break;
+        case PLUGIN_PCF:
+          currentState = GPIO_PCF_Read(gpioPort);
+          if (currentState==-1){
+            it->second.mode=PIN_MODE_OFFLINE;
+          } else if (it->second.mode==PIN_MODE_OFFLINE) {
+            it->second.mode=PIN_MODE_UNDEFINED;
+          }
+          eventString = F("PCF");
+          break;
+        default:
+          caseFound=false;
+        break;
+      }
+      if (caseFound && ((it->second.state != currentState) || (it->second.forceMonitor && it->second.monitor))) {
+        // update state
+        if (!it->second.task) it->second.state = currentState; //update state ONLY if task flag=false otherwise it will not be picked up by 10xSEC function
+        // send event
+        if (it->second.monitor) sendMonitorEvent(eventString.c_str(), gpioPort, currentState);
+      }
+      it->second.forceMonitor=0; //reset flag
+    }
+  }
+}
+
+// prefix should be either "GPIO", "PCF", "MCP"
+void sendMonitorEvent(const char* prefix, byte port, int8_t state)
+{
+  String eventString = prefix;
+  eventString += '#';
+  eventString += port;
+  eventString += '=';
+  eventString += state;
+  rulesProcessing(eventString);
+}
+
+bool checkValidPortRange(byte pluginID, byte port)
+{
+  bool returnValue = false;
+  switch (pluginID)
+  {
+    case PLUGIN_GPIO:
+      returnValue=(port>=0 && port<=PIN_D_MAX);
+    break;
+    case PLUGIN_MCP:
+    case PLUGIN_PCF:
+      returnValue=(port>=1 && port<=128);
+    break;
+  }
+  return returnValue;
+}
+
+void setGPIOPullupMode(byte port)
+{
+#if defined(ESP8266)
+  if (port == 16) {
+    pinMode(port, INPUT_PULLDOWN_16);
+  }
+  else {
+    pinMode(port, INPUT_PULLUP);
+  }
+#else // if defined(ESP8266)
+  pinMode(port, INPUT_PULLUP);
+#endif // if defined(ESP8266)
+}
```

Comment:
Just wondering, do the expanders have internal pull-up resistors?

Reply:
@giig1967g @TD-er PCF - no, MCP yes